### PR TITLE
Fixed #27068 -- Unified form field initial data retrieval.

### PR DESCRIPTION
**Pull request by jdufresne**
<https://code.djangoproject.com/ticket/27068>

### CODE REVIEW

File: django/forms/forms.py [hunk 1]
```diff
@@ -482,6 +482,12 @@ def visible_fields(self):
         """
         return [field for field in self if not field.is_hidden]
 
+    def get_initial(self, field_name, field):
```
Comment (timgraham) [inline django/forms/forms.py]: Add docstring? No need to document this method?

Comment (timgraham) [inline django/forms/forms.py]: I wonder if `get_initial_for_field` might make it more readable?

File: tests/forms_tests/tests/test_forms.py [hunk 2]
```diff
@@ -1904,6 +1904,28 @@ class UserRegistration(Form):
...
+            return 'John Doe'
+
+        class PersonForm(Form):
+            name = CharField(initial=initial_name, disabled=True)
+
+        # Assert base case without form data is correct.
+        form = PersonForm(auto_id=False)
+        self.assertHTMLEqual(
+            form.as_ul(),
+            '<li>Name: <input name="name" type="text" value="John Doe" disabled required /></li>',
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: Could you try to write these tests at a lower level, without the `as_ul()` output methods? They're difficult to debug now.

File: tests/forms_tests/tests/test_forms.py [hunk 3]
```diff
@@ -1904,6 +1904,28 @@ class UserRegistration(Form):
...
         )
 
+    def test_callable_initial_data_bound_form_disabled_field(self):
+        def initial_name():
+            return 'John Doe'
+
+        class PersonForm(Form):
+            name = CharField(initial=initial_name, disabled=True)
+
+        # Assert base case without form data is correct.
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: `# Without form data` seem sufficient.

File: tests/forms_tests/tests/test_forms.py [hunk 4]
```diff
@@ -1904,6 +1904,28 @@ class UserRegistration(Form):
...
+        form = PersonForm(auto_id=False)
+        self.assertHTMLEqual(
+            form.as_ul(),
+            '<li>Name: <input name="name" type="text" value="John Doe" disabled required /></li>',
+        )
+
+        # Again with form data. As the field is disabled, the output should
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: `# With form data. ...`

File: tests/forms_tests/tests/test_forms.py [hunk 5]
```diff
@@ -2021,6 +2043,96 @@ class DateTimeForm(Form):
...
+
+        class DateTimeForm(forms.Form):
+            dt = DateTimeField(initial=delayed_now, disabled=True)
+
+        form = DateTimeForm({})
+        self.assertEqual(form.errors, {})
+        self.assertEqual(form.cleaned_data, {'dt': now})
+
+    def test_datetime_changed_data_callable_with_microseconds(self):
+        now = datetime.datetime(2006, 10, 25, 14, 30, 45, 123456)
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: No need for intermediate variable?

File: tests/forms_tests/tests/test_forms.py [hunk 6]
```diff
@@ -2021,6 +2043,96 @@ class DateTimeForm(Form):
...
+
+        def delayed_now():
+            return now
+
+        class DateTimeForm(forms.Form):
+            dt = DateTimeField(initial=delayed_now, disabled=True)
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: You could use lambda to save some lines defining the function

**Comment by jdufresne** [comment]
Made adjustments based on feedback. Please review. Thanks.

File: django/forms/forms.py [hunk 7]
```diff
@@ -482,6 +482,17 @@ def visible_fields(self):
         """
         return [field for field in self if not field.is_hidden]
 
+    def get_initial_for_field(self, field, field_name):
+        """
+        Return initial data for field on form. Return the form's initial data
```
Comment (timgraham) [inline django/forms/forms.py]: For the second sentence, maybe: "Use initial data from the form or the field, in that order."

File: docs/ref/forms/api.txt [hunk 8]
```diff
@@ -248,6 +248,15 @@ precedence::
     <tr><th>Url:</th><td><input type="url" name="url" required /></td></tr>
     <tr><th>Comment:</th><td><input type="text" name="comment" required /></td></tr>
 
+.. method:: Form.get_initial_for_field(field, field_name)
+
+.. versionadded:: 1.11
+
+Use :meth:`~Form.get_initial_for_field()` to retrieve initial data for a field
```
Comment (timgraham) [inline docs/ref/forms/api.txt]: "for a form field"?

File: docs/ref/forms/api.txt [hunk 9]
```diff
@@ -248,6 +248,15 @@ precedence::
...
+on a form. This method handles data retrieval by first checking
```
Comment (timgraham) [inline docs/ref/forms/api.txt]: It retrieves data from Form.initial and Field.initial, in that order.

File: docs/releases/1.11.txt [hunk 10]
```diff
@@ -187,6 +187,10 @@ Forms
 * The new :attr:`CharField.empty_value <django.forms.CharField.empty_value>`
   attribute allows specifying the Python value to use to represent "empty".
 
+* The new :meth:`Form.get_initial_for_field()
+  <django.forms.Form.get_initial_for_field>` method allows retrieving initial
```
Comment (timgraham) [inline docs/releases/1.11.txt]: method returns initial data...

**Comment by timgraham** [comment]
Some tests for `get_initial_for_field()` might be nice.

**Comment by jdufresne** [comment]
Update text based on feedback and added recommended tests. All additional feedback is welcome, thanks!

**Pull request merged**